src_c/time.c

```diff
@@ -93,107 +90,125 @@ timer_callback(Uint32 interval, void *param)
+     * This clock uses the most accurate clock available on the system
+     * On most windows platforms, this has microsecond precision and on
+     * most Unix-based platforms (Linux, Mac, etc) this gives nanosecond
+     * precision (fallback to microsecond precision in rare cases)
+     * In worst cases (extremely rare) it may fall back to using 
+     * millisecond accuracy.
+     * Here, we divide by 1000 because we want ticks per millisecond, not
+     * ticks per second */
+    return SDL_GetPerformanceFrequency() / 1000;
 }
 
-static PyObject *
-time_get_ticks(PyObject *self)
-{
-    if (!SDL_WasInit(SDL_INIT_TIMER))
-        return PyInt_FromLong(0);
-    return PyInt_FromLong(SDL_GetTicks());
+#else /* IS_SDLv1 */
+static Uint64
+_pg_get_clock(void) {
+    return SDL_GetTicks();
 }
 
-static PyObject *
-time_delay(PyObject *self, PyObject *arg)
-{
-    int ticks;
-    PyObject *arg0;
-
-    /*for some reason PyArg_ParseTuple is puking on -1's! BLARG!*/
-    if (PyTuple_Size(arg) != 1)
-        return RAISE(PyExc_ValueError, "delay requires one integer argument");
-    arg0 = PyTuple_GET_ITEM(arg, 0);
-    if (!PyInt_Check(arg0))
-        return RAISE(PyExc_TypeError, "delay requires one integer argument");
-
-    ticks = PyInt_AsLong(arg0);
-    if (ticks < 0)
-        ticks = 0;
-
-    ticks = accurate_delay(ticks);
-    if (ticks == -1)
-        return NULL;
-    return PyInt_FromLong(ticks);
+static Uint64
+_pg_get_clock_precision(void) {
+    return 1;
 }
+#endif /* IS_SDLv1 */
 
-static PyObject *
-time_wait(PyObject *self, PyObject *arg)
+static double
+_pg_get_delta_millis(Uint64 start)
 {
-    int ticks, start;
-    PyObject *arg0;
+    return (double)(_pg_get_clock() - start) / _pg_get_clock_precision();
+}
 
-    /*for some reason PyArg_ParseTuple is puking on -1's! BLARG!*/
-    if (PyTuple_Size(arg) != 1)
-        return RAISE(PyExc_ValueError, "delay requires one integer argument");
-    arg0 = PyTuple_GET_ITEM(arg, 0);
-    if (!PyInt_Check(arg0))
-        return RAISE(PyExc_TypeError, "delay requires one integer argument");
+static Uint64
+_pg_get_ticks_from_millis(double delta_millis)
+{
+    return (Uint64)(delta_millis * _pg_get_clock_precision());
+}
 
-    if (!SDL_WasInit(SDL_INIT_TIMER)) {
-        if (SDL_InitSubSystem(SDL_INIT_TIMER)) {
-            return RAISE(pgExc_SDLError, SDL_GetError());
-        }
+static double
+_pg_accurate_delay(double millis)
+{
+    int sleeptime;
+    Uint64 endtime, starttime = _pg_get_clock();
+    
+    if (millis <= 0)
+        return 0.0;
+    
+    endtime = starttime + _pg_get_ticks_from_millis(millis);  
+    if (millis >= 20)
+        sleeptime = (int)millis - 5;
+    else if (millis >= 15)
+        sleeptime = (int)millis - 4;
+    else if (millis >= 10)
+        sleeptime = (int)millis - 3;
+    else if (millis >= 5)
+        sleeptime = (int)millis - 2;
+    else if (millis >= 3)
+        sleeptime = (int)millis - 1;
+    else
+        sleeptime = 0;
```

Comment:
What is this block of code doing? There is no comment and at first glance it seems fairly puzzling to make a delay for SDL Delay that is whatever was requested minus 5, if the delay is bigger than 20 milliseconds.

I'm sure it was written for a reason but to look at the reason is very unclear so a comment or perhaps moving this and the SDL_Delay() part below to a separate function with a descriptive name might help.

My uniformed guess is that the function is trying to use  SDL_Delay() to do the bulk of the delaying work because it is faster than a python while loop? But why different numbers have to be subtracted from this delay amount is a little bit lost on me so far.

Reply:
Yeah, I probably should have explained more with comments, I will add that soon.
My basic idea here is to get a good balance between a sleep and an accurate busy-loop delay, I try to do both, first by sleeping the processor for a while, and then doing a busy loop to cover for the remaining time. This way, the delay is accurate and the processor does not burn off by doing repeatedly busy loops.
That bit of code is just doing crude checks on the time delay to determine the time it has to sleep
If the delay is more than 20 millis, busy loop for 5-ish millis
If the delay is between 15 and 20 millis, busy loop for 4-ish millis and so on